### ISSUE TITLE
Translated elements can become blurry

### ISSUE THREAD

**Issue by jfrueh** [bcakmakoglu/revue-draggable#51]
After an element is dragged it can at times become blurry.  Sometimes it's very clear, other times it's not - it just depends on where it is positioned.  This root cause of this is the css translate property and this seems to affect most drag/drop libraries.

I think the best, most sure-fire fix is on dragend to use the top/left offsets to position the element rather than the translate property.  That was done in this library:  https://github.com/taye/interact.js/issues/285

I've attached an image showing the problem where the entire dark container is my draggable element.

Browser:  Chrome.

Thanks for considering a fix...

![image](https://user-images.githubusercontent.com/1854052/129135651-d943e17a-1ab6-4fdb-a928-d48bb09ebb0e.png)

**Comment by bcakmakoglu** [comment]
I'll take a look and will get back at you, thanks for opening the issue.

**Comment by bcakmakoglu** [comment]
@jfrueh 
I can't seem to reproduce the blurriness your image shows. I get some slightly fuzzy text but that's it.
Could you tell me what version of Chrome you're using and possibly show an excerpt of the html when the element gets blurry? 
What screen resolution does it happen at? Is it every screen size that produces the issue?

I implemented a minor fix that eventually produced this result but I'd like to further investigate to make sure this would actually solve the issue already (left is Firefox 91.0, right is Chrome 92.0.4515.131)
<img width="1238" alt="image" src="https://user-images.githubusercontent.com/78412429/129260744-bb3b050d-c560-4d97-a197-f0ff6ea44c5d.png">

**Comment by jfrueh** [comment]
Chrome version is 2.0.4515.131.  I also see this issue in Edge.  

I do NOT see the issue in FireFox so it seems to definitely be browser related.

I had a feeling this might be hard to repro.  On my side I'm using a 4k monitor at 3840 x 2160 resolution and I can reproduce this at any screen size so it doesn't seem to be related to the size of the browser window.  On some window sizes it's blurry all the time, on others it just depends on the objects position.  

If I manually set the position of the toolbar after drag (rather than using transform: translate) the blurriness seems to go away.  I could look at the code too but is removing transform after dragging is done and setting the position via 'fixed' or 'absolute' viable?  That seems to be the cross-browser fix that I'm seeing.

Thanks for considering...

**Comment by jfrueh** [comment]
...I could also look to handle this myself on the drag stop event by using el.getBoundingClientRect() to position the element and overwrite the transform property.  But would be better if this native to the plugin.

**Comment by bcakmakoglu** [comment]
I'm just concerned that using pos absolute/relative on the element could interfere with boundaries or an element's current position on the page, say if it's inside another box.

Can you try rounding the transformations to whole numbers? 
So if there's `transform: translate(10.0003px 10.0054px)` change it to 10px and 10px and see if the issue still occurs.
And also try applying `transform: rotate(-0.0000000001deg)` to it as well.

Let me know if that does anything.

**Comment by jfrueh** [comment]
I haven't seen anything in the transform that is not a whole number.  Maybe values are being rounded before they are set?  I think rounding to the nearest 1/2 pixel is what I've seen recommended.  The rotate transform didn't see to do anything either.  Sorry I can't get you an exact repro of this - makes it really hard for you.

Yes, using absolute could interfere if the other box didn't have a relative position set so that would be a requirement.    

On drag end I can look to set the position manually if you think that is the best route.

**Comment by bcakmakoglu** [comment]
@jfrueh 
Well until I have an actual fix that wouldn't interfere with existing positions and actually fix the chrome blurriness I'd have to ask you to set it yourself on the stop event.
__But__ I will be looking into providing a solution. I could possibly add an option to set the fix if necessary, so it wouldn't be a default setting in case someone does want to use a fixed/absolute position on the element itself.

**Comment by jfrueh** [comment]
I think making that fix optional sounds like good plan.  Thank you.

**Comment by bcakmakoglu** [comment]
In the 3.0.0-0 Version there's an option called `enableTransformFix` which you can use now.
Try it out and let me know if it works as desired.
You can either pass a boolean or { position: 'relative' | 'absolute' }.
It will use a relative position by default.

```vue
<div v-draggable="{ enableTransformFix: true }" />
```

**Comment by jfrueh** [comment]
3.0.0-0 isn't working the same as 2.x was for me. 

In the console I'm now seeing this:   [Vue warn]: You may have an infinite update loop in a component render function.

I'm having trouble tracking down the source of this error but it's being caused by the upgrade to 3.0.0-0.  I'm calling v-draggable this way:

`  <div
    v-if="dataReady"
    id="floatingToolbar"
    ref="foo"
    v-draggable="{ handle: '#toolbarDrag', disabled: $vuetify.breakpoint.mobile, enableTransformFix:  true }"
    :class="classes"
    :style="style"
    @start="dragStart"
    @stop="dragStop"
  />`

It might be important to mention that my 'style' variable looks like this:

`style: {
        left: '0px',
        bottom: '0px',
        width: 'initial',
        top: 'initial'
      },`

I wonder if with this version there is now some issue with me setting top, left, bottom values on the draggable element itself? (these values are currently being set by me to position the draggable element wherever needed based on a different element being clicked)....although I did comment this out and the error was still present.

Any thoughts on where to look first?

**Comment by bcakmakoglu** [comment]
I don't think it's your stylings that are an issue but the directive constantly updating on the watchEffect hook.
Try the 3.0.0-1 release and let me know if the issue persists, in my vue2 tests it worked without an issue.

**Comment by jfrueh** [comment]
I still have the same error.  ("You may have an infinite update loop in a component render function.")

When I click and drag the element jumps to a random spot as well...

**Comment by bcakmakoglu** [comment]
3.0.0.-2 should fix that infinite loop finally 😅

The jumping seems strange, I can't seem to reproduce it using your example and applying the style you used on the element.
<img width="502" alt="image" src="https://user-images.githubusercontent.com/78412429/129608394-57cb1a83-65e9-4a64-871f-4f08efb44c87.png">
Results in normal dragging behavior with no jumps whatsoever.

Can you try it with the 3.0.0-2 release and let me know your results? 🏌️‍♀️

**Comment by jfrueh** [comment]
Infinite loop is fixed.  The jumping is still present, but only when  enableTransformFix: true is set.  Can you confirm if on your tests you had enableTransformFix set to true?

**Comment by bcakmakoglu** [comment]
Yes, I had it set to true, otherwise the snippet above wouldn't have had the position top & left set :) I didn't add any other handler functions to the element.
Are you modifying the positions on drag start or drag end manually?

**Comment by jfrueh** [comment]
So it looks like the "jumping" issue comes up only after I execute a reset position on the draggable element.

I've tried to reset the position using 2 methods:
1.  this.controlledPosition = { x: 0, y: 0 };  (the controlled position method)
2.  this.$refs.foo['revue-draggable'].value = { ...this.$refs.foo['revue-draggable'].value, x: 0, y: 0 };

Prior to executing either of these methods it all works perfect.  After either of these it becomes unpredictable.

Thank you for the help.

**Comment by bcakmakoglu** [comment]
Alright, I'll check out what goes wrong an will update you asap :)

**Comment by bcakmakoglu** [comment]
Should be fixed in 3.0.0-3, issue seems to be a missing transformation after the state has been overwritten.
It should now transform as soon as you change the elements position or in your case apply a position top & left if the transform fix is enabled.

Let me know if that solved your jumping issue.

**Comment by jfrueh** [comment]
Still seeing the same issue, but have narrowed it down to being due to the :style binding on the draggable element.  

Here is some code you can use:

`  <div id="el" ref="foo" v-draggable="{ handle: '#toolbarDrags', disabled: $vuetify.breakpoint.mobile, enableTransformFix: true }" style="width:600px;height:200px;background-color:green;display:flex;position:relative" :style=style" > <div id="toolbarDrags" style="width:30px;height:30px;background-color:black;" ></div> </div>`

The style object that this is bound to looks like this:
`var style = {
        left: '500px;',
        bottom: 'initial',
        width: 'initial',
        top: '500px'
      };`

I reset this style when other elements on the page are clicked.

So to repro take an action that updates the style object, drag the element around, take another action to update the style variable, and try to drag around again.  The draggable should jump out of position.

Thank you for taking a look.

**Comment by bcakmakoglu** [comment]
@jfrueh 
I'm assuming that setting the style and then letting Draggable re-set the style is what's causing the issue as Draggable doesn't know what position you used for the styling yourself but I haven't checked it out yet but I will asap but today I'm quite busy, please excuse the delay and thanks for your patience in fixing this and helping me out 😄

**Comment by bcakmakoglu** [comment]
@jfrueh 
Yeah it's like I assumed; Using the transform fix it's not really a good idea to also try to override the elements position by adding a styling.
You're basically interfering with Draggable which keeps track of the x, y positions based on mouse events and not based on your styles you provide.
So whenever you start dragging Draggable applies it's tracked values, which differ from the ones you're setting and you experience "jumping".

The current way to apply a starting/default position is to set it as as the option "defaultPosition" i.e.
```
<div v-draggable="{ defaultPosition: { x: 500, y: 500 } }" >Drag me</div>
```
Though in 3.0.0-3 the default position might be broken with the transformation fix but it is working in 3.0.0-4.
In the new release the style applied beforehand `<div style="top: 500px; left: 500px" v-draggable>Drag me</div>` will be used as the default position. 

As for resetting the position afterwards you will have to either override the current x and y values using the ref and state override or simply take control of the element and provide the x and y values yourself.

So in short:
* Add default positions by passing the appropriate option `defaultPosition: { x: number; y: number; }` or let `Draggable` handle setting the default positions using styles
* Reset positions by either using the state ref 
* or reset positions by taking over `Draggable` and providing positions by passing `position: { x: number; y: number; }`

See if this finally solves your issue 😄

**Comment by bcakmakoglu** [comment]
@jfrueh any updates?

**Comment by jfrueh** [comment]
Hey - been tied up with a few other things and haven't had a chance to get back to this yet.  I'm going to close this out with this comment and if I come across any issues will just re-open in the future.  Appreciate the follow-up and help!
